ibc-core/ics02-client/context/src/context.rs

```diff
@@ -102,3 +103,62 @@ pub trait ClientExecutionContext:
         height: Height,
     ) -> Result<(), ContextError>;
 }
+
+/// An optional trait that extends the client validation context capabilities by
+/// providing additional methods for validating a client state, particularly
+/// benefiting ICS-07 Tendermint clients by granting access to essential
+/// information from hosts.
+///
+/// Categorized under ICS-02, as it may also be utilized by other types of light
+/// clients. Developers can view this trait as an example of a custom context
+/// definition that expands client validation capabilities, according to their
+/// specific light client requirements.
+pub trait ExtClientValidationContext: ClientValidationContext {
+    /// Returns the current timestamp of the local chain.
+    fn host_timestamp(&self) -> Result<Timestamp, ContextError>;
+
+    /// Returns the current height of the local chain.
+    fn host_height(&self) -> Result<Height, ContextError>;
+
+    /// Returns all the heights at which a consensus state is stored.
+    fn consensus_state_heights(&self, client_id: &ClientId) -> Result<Vec<Height>, ContextError>;
+
+    /// Search for the lowest consensus state higher than `height`.
+    fn next_consensus_state(
+        &self,
+        client_id: &ClientId,
+        height: &Height,
+    ) -> Result<Option<Self::ConsensusStateRef>, ContextError>;
+
+    /// Search for the highest consensus state lower than `height`.
+    fn prev_consensus_state(
+        &self,
+        client_id: &ClientId,
+        height: &Height,
+    ) -> Result<Option<Self::ConsensusStateRef>, ContextError>;
+}
+
+/// An optional trait that extends the client context required during execution.
+///
+/// This trait as it stands right now serves as a trait alias for types that
+/// implement both [`ExtClientValidationContext`] and
+/// [`ClientExecutionContext`], and it is automatically implemented for such
+/// types.
+///
+/// Light client developers who wish to define and utilize their own custom
+/// client contexts may choose to introduce execution methods within a similar
+/// trait, which would allow them to store additional client-specific data and
+/// conduct execution operations more efficiently, tailored to their light
+/// client implementation.
+pub trait ExtClientExecutionContext: ExtClientValidationContext + ClientExecutionContext {}
+
+impl<T> ExtClientExecutionContext for T where T: ExtClientValidationContext + ClientExecutionContext {}
+
+/// General-purpose helper converter enabling `TryInto` and `From` conversions
+/// primarily intended between an enum and its variants. This usually used by
+/// standalone functions as a trait bound allowing them to obtain the concrete
+/// local type from the enum containing that concrete type as its variant like
+/// when enum `AnyConsensusState` contains the Tendermint `ConsensusState`.
+pub trait TypeCaster<C, E>: TryInto<C, Error = E> + From<C> {}
```

Comment:
Here are my suggestions for renaming: `Convertible` or `BiConvertible`.

Reply:
[274e0b1](https://github.com/cosmos/ibc-rs/pull/1166/commits/274e0b1c2e67f8a7f68ff23f1b3ce04cc5b16961)